### PR TITLE
Add parsing support for linestring constructor

### PR DESCRIPTION
**Pull request by Ayman161803**
Signed-off-by: Ayman161803 <mohammedaymannawaz@gmail.com>

## Description
This PR adds parsing support to ``LINESTRING( POINT(x,y), POINT(x,y) ... )`` constructor

Note: Linestring takes in more than one point as argument. Test cases have been altered accordingly. Parsing ``Linestring(Point(x,y))`` with just one point argument gives the following error

image

## Related Issue(s)
#8604 

## Checklist

-   [x] "Backport to:" labels have been added if this change should be back-ported
-   [x] Tests were added or are not required
-   [x] Documentation was added or is not required

### CODE REVIEW

File: go/vt/sqlparser/ast.go [hunk 1]
```diff
@@ -2710,6 +2710,11 @@ type (
 		YCordinate Expr
 	}
 
+	//LineString represents LineString(POINT(x,y), POINT(x,y), ..) expression
+	LineStringExpr struct {
+		PointParams []*PointExpr
```
Comment (systay) [inline go/vt/sqlparser/ast.go:2715]: A `LineString()` function only takes point values as arguments, but these could come from various sources. You could use column values, user defined variables, or other functions that return `point` values. I don't think we can hard code these arguments to `PointExpr`

```sql
mysql> create table geom (pt1 point, pt2 point);
Query OK, 0 rows affected (0.01 sec)

mysql> select linestring(pt1, pt2) from geom;
```

☝️ this is a small example showing what I mean.

My suggestion is to use `Expr` instead of `*PointExpr`

**Comment by systay** [comment]
The error you are seeing for a single point value is not a syntax error, and should not be stopped by the parser. I'd suggest letting the parser accept it, and MySQL or evalengine will fail it with the proper error message

File: go/vt/sqlparser/ast.go (hunk 1 above)
Comment (Ayman161803) [reply go/vt/sqlparser/ast.go]: That makes a lot of sense. I have updated yacc grammar to accept ``expression_list`` instead of ``point_expression_list``, adding a test case to show that the example that u gave gets parsed now.

**Comment by Ayman161803** [comment]
> The error you are seeing for a single point value is not a syntax error, and should not be stopped by the parser. I'd suggest letting the parser accept it, and MySQL or evalengine will fail it with the proper error message

I have altered ``sql.y`` taking this into consideration. The ``Linestring`` constructor will now accept just one point argument too as shown in [this](https://github.com/vitessio/vitess/blob/3a522ff039b706438fbeb6a67ef76fee19fcc9e8/go/vt/sqlparser/parse_test.go#L188) test case

**Review by GuptaManan100 — CHANGES_REQUESTED** [review]
I checked the MySQL docs, and it does seem that having a lineString with a single point value is allowed by the parser, albeit an execution error - 
```MySQL
mysql [localhost:8026] {msandbox} (test) > create table x1(location GEOMETRYCOLLECTION DEFAULT (LINESTRING(POINT(4, 5))));
Query OK, 0 rows affected (0.24 sec)

mysql [localhost:8026] {msandbox} (test) > select linestring(point(1,1));
ERROR 3037 (22023): Invalid GIS data provided to function linestring.
```

In Vitess too, we should allow line string to have a single point.

File: go/vt/sqlparser/testdata/select_cases.txt [hunk 2]
```diff
@@ -9491,10 +9491,10 @@ OUTPUT
 select count(test.t1.b) from t1
 END
 INPUT
-select st_difference((convert(st_polygonfromwkb(linestring(point(1,1))) using gb18030)), st_geomcollfromwkb(point(1,1)));
+select st_difference((convert(st_polygonfromwkb(linestring(point(1,1), point(2,2))) using gb18030)), st_geomcollfromwkb(point(1,1)));
 END
 OUTPUT
-select st_difference(convert(st_polygonfromwkb(linestring(point(1, 1))) using gb18030), st_geomcollfromwkb(point(1, 1))) from dual
```
Comment (GuptaManan100) [inline go/vt/sqlparser/testdata/select_cases.txt:9497]: Nevermind, it is already accepting single values. Just revert this change, so that we have a test for that too.

File: go/vt/sqlparser/ast_format.go [hunk 3]
```diff
@@ -2759,3 +2759,16 @@ func (node *Variable) Format(buf *TrackedBuffer) {
...
+// Format formats the node.
+func (node *LineStringExpr) Format(buf *TrackedBuffer) {
+	buf.literal("linestring(")
+	if len(node.PointParams) > 0 {
+		var prefix string
+		for _, n := range node.PointParams {
+			buf.astPrintf(node, "%s%v", prefix, n)
+			prefix = ", "
+		}
+	}
```
Comment (GuptaManan100) [inline go/vt/sqlparser/ast_format.go:2772]: The code on how to print a list of Expressions is already present in the format function of `Exprs`. So we can simply use - 
```golang
// Format formats the node.
func (node *LineStringExpr) Format(buf *TrackedBuffer) {
	buf.astPrintf(node, "linestring(%v)", node.PointParams)
}
```

File: go/vt/sqlparser/testdata/select_cases.txt (hunk 2 above)
Comment (Ayman161803) [reply go/vt/sqlparser/testdata/select_cases.txt]: Reverted

File: go/vt/sqlparser/ast_format.go (hunk 3 above)
Comment (Ayman161803) [reply go/vt/sqlparser/ast_format.go]: Updated ``Format`` function accordingly

**Review by GuptaManan100 — APPROVED** [review]
LGTM now! Good work!

**Review by systay — APPROVED** [review]
